tech-spec.md

```diff
@@ -0,0 +1,310 @@
+    describeTableQuery(con, rules, table, query)
+    extractFilteredData(con, table, query)
+```
+
+## Rule schema parsing
+
+1. open schema file
+2. read lines
+3. parse each line into a rule obj
+4. add each rule obj to a dictionary with rule id as key
+
+## AST generation
+
+Node kinds:
+- **root**: AST root
+- **share**: what to share with each org
+- **table**: table name
+- **select**: lists column name values
+- **group**: groups filters together
+- **filter**: defines a filter with operator, key and value
+- **key**: field name
+- **value**: generic
+
+Node structure:
+
+- (ruleId: int)
+- kind: NodeKind
+- value: str
+- children: List[Node]
+
+For each rule, a node is added based on its mode:
+- select:
+    - (**select**, ("" or "all")):
+        - (**value**, column) for column in rule.key
+- filter:
+    - (**filter**, rule.operator):
+        - (**key**, rule.key)
+        - (**value**, x) for x in rule.value
+- group:
+    - (**group**, rule.operator):
+        - nodes where node.ruleId in rule.value
+- share:
+    - (**root**, ""):
+        - (**share**, org) for org in rule.key
+            - (**table**, x) for x in select-rule-tables
+                - filter-root node
+
+Example rules with its generated tree:
+
+```csv
+ruleId,table,mode,key,operator,value,notes
+10,measures,select,NA,NA,all,
+11,measures,select,NA,NA,measure;value;unit;aggregation,
+12,measures,filter,measure,=,mPox,
+13,measures,filter,reportDate,in,2021-01-01;2021-12-31,
+14,NA,group,NA,AND,12;13,
+15,measures,filter,measure,=,cov,
+16,measures,filter,reportDate,>=,2020-01-01,
+17,NA,group,NA,AND,15;16,
+18,NA,group,NA,OR,14;17,
+19,NA,share,ohri,NA,11;18,
+20,NA,share,other,NA,10,
+```
+
+```
+(root, "")
+    (share, "OHRI")
+        (table, "measures")
+            (select, "")
+                (value, "measure")
+                (value, "value")
+                (value, "unit")
+                (value, "aggregation")
+            (group, "OR")
+                (group, "AND")
+                    (filter, "=")
+                        (key, "measure")
+                        (value, "mPox")
+                    (filter, "in")
+                        (key, "reportDate")
+                        (value, "2021-01-01")
+                        (value, "2021-12-31")
+                (group, "AND")
+                    (filter, "=")
+                        (key, "measure")
+                        (value, "cov")
+                    (filter, ">=")
+                        (key, "reportDate")
+                        (value, "2020-01-01")
+    (share, "other")
+        (table, "measures")
+            (select, "all")
+```
+
+## SQL query generation
+
+SQL queries are (recursively) generated from each table node of the AST. Values
+are separated from the query to prevent injections. Multiple kinds of queries
+can be generated for different purposes, including only getting the row-counts
+or column names.
+
+- **table**:
```

Comment:
Would it be possible to write this as human readable pseudo code? Its hard to read/understand as it is. If it will take a lot of time then don't worry about it.

Reply:
done